onecodex/stats.py

```diff
@@ -0,0 +1,158 @@
+            whether or not they are significantly different. See `skbio.stats.composition.ancom` for
+            details.
+
+            `percentile_df`: `pd.DataFrame` containing percentile abundances of taxa in each group.
+            See `skbio.stats.composition.ancom` for details.
+
+            `chart`: Faceted Altair boxplot of significantly different taxa and their abundances in
+            each group. Will be `None` if no significantly different taxa are detected.
+
+        See Also
+        --------
+        skbio.stats.composition.ancom
+        skbio.stats.composition.multiplicative_replacement
+
+        """
+        import altair as alt
+        import pandas as pd
+        from skbio.stats.composition import ancom, multiplicative_replacement
+
+        if not isinstance(group_by, tuple):
+            group_by = (group_by,)
+
+        # Munge the metadata first in case there's any errors
+        metadata_df, magic_fields = self._metadata_fetch([group_by])
+
+        # Normalize because `multiplicative_replacement()` requires compositions
+        taxa_df = self.to_df(rank=rank, normalize=True)
+        taxa_df = taxa_df.rename(
+            columns=lambda tax_id: f"{self.taxonomy['name'][tax_id]} ({tax_id})"
+            if tax_id in self.taxonomy["name"]
+            else tax_id
+        )
+
+        # `multiplicative_replacement()` will error on rows that are all zeros
+        prev_index = taxa_df.index
+        taxa_df = taxa_df.loc[(taxa_df != 0).any(axis=1)]
+        num_dropped = len(prev_index) - len(taxa_df.index)
+        if num_dropped > 0:
+            warnings.warn(
+                f"{num_dropped} sample{'s' if num_dropped > 1 else ''} with zero abundance across "
+                f"all taxa {'were' if num_dropped > 1 else 'was'} ignored.",
+                StatsWarning,
+            )
+
+        # Drop any samples from the metadata df that are not in the taxa df
+        metadata_df = metadata_df.filter(items=taxa_df.index, axis="index")
+
+        num_groups = metadata_df[magic_fields[group_by]].nunique(dropna=False)
+        if num_groups < 2:
+            raise StatsException("`group_by` must have at least two groups to test between.")
+        if num_groups == len(metadata_df):
+            raise StatsException("Each group defined by `group_by` contains only one value.")
+
+        # Need to use pseudocounts or `multiplicative_replacement()` because ANCOM can't handle
+        # zeros
+        multiplicative_replacement_df = pd.DataFrame(
+            multiplicative_replacement(taxa_df), columns=taxa_df.columns, index=taxa_df.index
+        )
+        ancom_df, percentile_df = ancom(
+            multiplicative_replacement_df,
+            metadata_df[magic_fields[group_by]],
+            alpha=alpha,
+            multiple_comparisons_correction="holm-bonferroni",
+        )
+
+        chart = None
+        reject_column = ancom_df["Reject null hypothesis"]
+        if reject_column.any():
```

Comment:
I might move the plotting into a `plot()` method on `DifferentialAbundanceResults`, that way users can opt into plotting and that logic/interface is kept separate from the stats test. The plotting code relies on several pieces of intermediate computation (e.g. `taxa_df`, `metadata_df`, etc.), so it probably doesn't make sense to have a separate `SampleCollection.plot_differential_abundance()` method that would have to recompute all that data or accept it as input. A plot method on the dataclass seems cleaner.